go/fn/object_test.go

```diff
@@ -10,6 +10,106 @@ import (
+apiVersion: apps/v1
+kind: Service
+metadata:
+  name: my-config
+spec:
+  volumeClaimTemplates:
+    - metadata:
+        labels:
+          testkey: testvalue
+`)
+	resource, _ := ParseKubeObject(input)
+	resourceNoGroup, _ := ParseKubeObject(inputNoGroup)
+	resourceDiffKind, _ := ParseKubeObject(inputDifferentKind)
+	testCases := map[string]struct {
+		resource         *KubeObject
+		resourceNoGroup  *KubeObject
+		resourceDiffKind *KubeObject
+		group            string
+		version          string
+		kind             string
+		result           bool
+		resultNoGroup    bool
+		resultDiffKind   bool
+	}{
+		"IsGVK provided with no version, should match all versions": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "apps", version: "", kind: "StatefulSet",
+			result: true, resultNoGroup: true, resultDiffKind: false,
+		},
+		"IsGVK provided with no group, should match all groups": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "", version: "v3", kind: "StatefulSet",
+			result: true, resultNoGroup: true, resultDiffKind: false,
+		},
+		"IsGVK provided with no kind, should match all kinds": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "apps", version: "v3", kind: "",
+			result: true, resultNoGroup: true, resultDiffKind: false,
+		},
+		"IsGVK provided with no fields, should match all resources": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "", version: "", kind: "",
+			result: true, resultNoGroup: true, resultDiffKind: true,
+		},
+		"IsGVK provided with all fields, can only match if the resource has the same field value or the field is nil": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "apps", version: "v3", kind: "StatefulSet",
+			result: true, resultNoGroup: true, resultDiffKind: false,
+		},
+		"IsGVK provided with only kind, can only match if the kind is the same or kind is nil": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "", version: "", kind: "StatefulSet",
+			result: true, resultNoGroup: true, resultDiffKind: false,
+		},
+		"IsGVK provided with only group, can only match if the group is the same or group is nil": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "appWrong", version: "", kind: "",
+			result: false, resultNoGroup: true, resultDiffKind: false,
+		},
+		"IsGVK provided with only version, can only match if the version is the same or version is nil": {
+			resource: resource, resourceNoGroup: resourceNoGroup, resourceDiffKind: resourceDiffKind,
+			group: "", version: "v1", kind: "",
+			result: false, resultNoGroup: false, resultDiffKind: true,
+		},
+	}
+
+	for testName, data := range testCases {
```

Comment:
nit: use [go subtests](https://go.dev/blog/subtests) (see the section titled "Table-driven tests using subtests")

This would look like: 

```
    for testName, data := range testCases {
        t.Run(testName + "/" + resource, func(t *testing.T) {
   		    assert.Equal(t, resource.IsGVK(data.group, data.version, data.kind), data.result)
        })
        t.Run(testName + "/" + resourceNoGroup, func(t *testing.T) {
   		    assert.Equal(t, resourceNoGroup.IsGVK(data.group, data.version, data.kind), data.resultNoGroup)
        })
        t.Run(testName + "/" + resourceDiffKind, func(t *testing.T) {
   		    assert.Equal(t, resourceDiffKind.IsGVK(data.group, data.version, data.kind), data.resultDiffKind)
        })
	}
```

In the case of a test failure, subtests are very useful and will give you some nice formatted test output to tell you which specific subtest failed, without needing to provide your own custom message.

Reply:
Thanks a lot! Very useful suggestions, now I can run each case separately.